File: typescript/src/auth/CoinmateAuth.ts
import * as crypto from 'crypto';

/**
 * Authentication utility for Coinmate API
 * Implements HMAC-SHA256 signature generation
 */
export class CoinmateAuth {
  constructor(
    private readonly clientId: string,
    private readonly publicKey: string,
    private readonly privateKey: string
  ) {}

  /**
   * Generate a nonce (unix timestamp in milliseconds)
   * Each request must have a nonce greater than the previous one
   */
  generateNonce(): number {
    return Date.now();
  }

  /**
   * Create HMAC-SHA256 signature for API authentication
   * Signature input: nonce + clientId + publicKey
   *
   * @param nonce The nonce value (must be unique and increasing)
   * @returns Uppercase hexadecimal signature string
   */
  createSignature(nonce: number): string {
    const message = `${nonce}${this.clientId}${this.publicKey}`;

    const hmac = crypto.createHmac('sha256', this.privateKey);
    hmac.update(message);

    return hmac.digest('hex').toUpperCase();
  }

  /**
   * Create authentication parameters for a request
   */
  getAuthParams(): {
    clientId: string;
    publicKey: string;
    nonce: string;
    signature: string;
  } {
    const nonce = this.generateNonce();
    const signature = this.createSignature(nonce);

    return {
      clientId: this.clientId,
      publicKey: this.publicKey,
      nonce: nonce.toString(),
      signature,
    };
  }

  getClientId(): string {
    return this.clientId;
  }

  getPublicKey(): string {
    return this.publicKey;
  }
}

File: python/src/auth/coinmate_auth.py
"""HMAC-SHA256 authentication for Coinmate API"""
import hmac
import hashlib
import time


class CoinmateAuth:
    """Handles authentication for Coinmate API requests"""

    def __init__(self, client_id: str, public_key: str, private_key: str):
        """
        Initialize authentication handler

        Args:
            client_id: Coinmate client ID
            public_key: Coinmate public API key
            private_key: Coinmate private API key
        """
        self.client_id = client_id
        self.public_key = public_key
        self.private_key = private_key

    def get_nonce(self) -> int:
        """
        Generate nonce (current timestamp in milliseconds)

        Returns:
            Current timestamp in milliseconds
        """
        return int(time.time() * 1000)

    def create_signature(self, nonce: int) -> str:
        """
        Create HMAC-SHA256 signature

        Args:
            nonce: Timestamp in milliseconds

        Returns:
            Uppercase hex signature
        """
        message = f"{nonce}{self.client_id}{self.public_key}"
        signature = hmac.new(
            self.private_key.encode('utf-8'),
            message.encode('utf-8'),
            hashlib.sha256
        ).hexdigest().upper()
        return signature

    def get_auth_params(self) -> dict:
        """
        Get authentication parameters for API request

        Returns:
            Dictionary with clientId, publicKey, nonce, and signature
        """
        nonce = self.get_nonce()
        signature = self.create_signature(nonce)

        return {
            'clientId': self.client_id,
            'publicKey': self.public_key,
            'nonce': str(nonce),
            'signature': signature
        }
